server/tserver/src/main/java/org/apache/accumulo/tserver/log/RecoveryLogsIterator.java

```diff
@@ -19,61 +19,87 @@
 package org.apache.accumulo.tserver.log;
 
 import java.io.IOException;
+import java.util.AbstractMap;
 import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map.Entry;
 
+import org.apache.accumulo.core.client.Scanner;
+import org.apache.accumulo.core.client.ScannerBase;
+import org.apache.accumulo.core.client.rfile.RFile;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.server.ServerContext;
 import org.apache.accumulo.server.fs.VolumeManager;
+import org.apache.accumulo.server.log.SortedLogState;
+import org.apache.accumulo.tserver.logger.LogEvents;
 import org.apache.accumulo.tserver.logger.LogFileKey;
 import org.apache.accumulo.tserver.logger.LogFileValue;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.Text;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.google.common.collect.Iterators;
-import com.google.common.collect.UnmodifiableIterator;
 
 /**
  * Iterates over multiple sorted recovery logs merging them into a single sorted stream.
  */
-public class RecoveryLogsIterator implements CloseableIterator<Entry<LogFileKey,LogFileValue>> {
+public class RecoveryLogsIterator
+    implements Iterator<Entry<LogFileKey,LogFileValue>>, AutoCloseable {
 
   private static final Logger LOG = LoggerFactory.getLogger(RecoveryLogsIterator.class);
 
-  List<CloseableIterator<Entry<LogFileKey,LogFileValue>>> iterators;
-  private UnmodifiableIterator<Entry<LogFileKey,LogFileValue>> iter;
+  private final List<Scanner> scanners;
+  private final Iterator<Entry<Key,Value>> iter;
 
   /**
-   * Iterates only over keys in the range [start,end].
+   * Scans the files in each recoveryLogDir over the range [start,end].
    */
-  RecoveryLogsIterator(VolumeManager fs, List<Path> recoveryLogPaths, LogFileKey start,
-      LogFileKey end) throws IOException {
+  RecoveryLogsIterator(ServerContext context, List<Path> recoveryLogDirs, LogFileKey start,
+      LogFileKey end, boolean checkFirstKey, LogEvents... colFamToFetch) throws IOException {
```

Comment:
Each method called during the recovery process sends different columns to fetch:
findMaxTabletId -> OPEN
findRecoverySeq -> COMPACTION_START, COMPACTION_FINISH
playbackMutations -> MUTATION, MANY_MUTATIONS

Even though the first byte of each row is the event type, I thought it made a difference by not getting everything each time. For example, for `findMaxTabletId` without fetching any columns, wouldn't the Scanner get everything in the part of the WAL read? Where as, calling fetch on the OPEN column families will only return the rows that define tablet? Or does setting the Range override any columns fetched since it won't scan beyond the Range? If so, I should stop storing the event type, that would save some disk space.

Reply:
I think you are right cause it won't matter once seek is called and we seek to the range.

Reply:
> I think you are right cause it won't matter once seek is called and we seek to the range.

Maybe, it comes down to analyzing the events passed in and how those correspond to the byte for each call of the method.

> If so, I should stop storing the event type, that would save some disk space.

Multiple event types map to a single byte, so the byte may not contain enough information to reconstruct the exact event type.  May still need to store it somewhere.

Reply:
> Multiple event types map to a single byte, so the byte may not contain enough information to reconstruct the exact event type. May still need to store it somewhere.

I forgot we do need the specific event type to reconstruct the Event in `fromKey()`
`logFileKey.event = LogEvents.valueOf(key.getColumnFamilyData().toString());`

Reply:
Dropped colFamToFetch in e831472